plugins/ccgx/fu-ccgx-native-hid-device.c

```diff
@@ -0,0 +1,505 @@
+	//			    "app type mismatch, expected 0x%x, got 0x%x",
+	//			    self->fw_app_type,
+	//			    fw_app_type);
+	//		return NULL;
+	//	}
+	//}
+	fw_mode = fu_ccgx_firmware_get_fw_mode(FU_CCGX_FIRMWARE(firmware));
+	if (fw_mode != fu_ccgx_fw_mode_get_alternate(self->fw_mode)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "FuCcgxFwMode mismatch, expected %s, got %s",
+			    fu_ccgx_fw_mode_to_string(fu_ccgx_fw_mode_get_alternate(self->fw_mode)),
+			    fu_ccgx_fw_mode_to_string(fw_mode));
+		return NULL;
+	}
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_ccgx_native_hid_device_write_firmware(FuDevice *device,
+					 FuFirmware *firmware,
+					 FuProgress *progress,
+					 FwupdInstallFlags flags,
+					 GError **error)
+{
+	FuCcgxNativeHidDevice *self = FU_CCGX_NATIVE_HID_DEVICE(device);
+
+	g_debug("Operating Mode:  0x%02x", self->fw_mode);
+	switch (self->fw_mode) {
+	case FU_CCGX_FW_MODE_BOOT:
+	case FU_CCGX_FW_MODE_FW2:
+		// Update Image 1
+		g_debug("Flashing Image 1");
+		if (!fu_ccgx_native_hid_flash_firmware_image(self, firmware, progress, 1, error))
+			return FALSE;
+
+		g_debug("Add wait for replug");
+		fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+		fu_device_add_private_flag(device, FU_CCGX_NATIVE_HID_DEVICE_IS_IN_RESTART);
+		break;
+	case FU_CCGX_FW_MODE_FW1:
+		// Update Image 2
+		g_debug("Flashing Image 2");
+		if (!fu_ccgx_native_hid_flash_firmware_image(self, firmware, progress, 2, error))
+			return FALSE;
+
+		g_debug("Add wait for replug");
+		fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+		fu_device_add_private_flag(device, FU_CCGX_NATIVE_HID_DEVICE_IS_IN_RESTART);
+		break;
+	default:
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+// TODO: Need this? Why are there two?
+static void
+fu_ccgx_native_hid_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 5, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 45, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 5, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 45, "reload");
+}
+
+static void
+fu_ccgx_native_hid_device_init(FuCcgxNativeHidDevice *self)
+{
+	// FIXME: I don't think this is try anymore
+	fu_device_add_protocol(FU_DEVICE(self), "com.infineon.ccgx");
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UNSIGNED_PAYLOAD);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_DUAL_IMAGE);
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_INTEL_ME2);
+	fu_device_add_internal_flag(FU_DEVICE(self), FU_DEVICE_INTERNAL_FLAG_ONLY_WAIT_FOR_REPLUG);
+	fu_device_add_internal_flag(FU_DEVICE(self), FU_DEVICE_INTERNAL_FLAG_REPLUG_MATCH_GUID);
```

Comment:
do we need this flag?

Reply:
Yes, because the firmware mode changes between fw1 and fw2.
Otherwise when the device comes back after the update fwupd won't recognize that it's the same.

I'm not quite sure how it figures out that it's the same device. Because the "device id" does change. But it seems to work.

Reply:
Ahh of course, you're 100% right to use the GUID to identify the "same" device.